### PR TITLE
Metadata version validation

### PR DESCRIPTION
**Pull request by omerKarkKatz**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/41264

## Description
This was a dev issue about the meta data schema. to be fixed. troubleshooted the problem to come from the LooseVersion module. that was comparing to different version types. added a check in the _is_pack_meta_file_structure_valid to validate the version type in the meta data. if the version is not OK will not proceed to check the version bump where previously was raised an exception. 

## Screenshots
Paste here any images that will help the reviewer

## Must have
- [ ] Tests
- [ ] Documentation
- [ ] [VS Code Extension](https://docs.gitlab.com/ee/ci/yaml/#tags)
  - [ ] Functionality implemented in the extension - <u>\<link-to-pr-here\></u>
  - [ ] N/A - Functionality cannot be implemented in the extension because <u>\<add-reason-here\></u>

### CODE REVIEW

File: demisto_sdk/commands/common/hook_validations/pack_unique_files.py [hunk 1]
```diff
@@ -308,11 +307,10 @@ def validate_version_bump(self):
...
         if LooseVersion(old_version) < LooseVersion(current_version):
             return True
-
-        elif self._add_error(Errors.pack_metadata_version_should_be_raised(self.pack, old_version), metadata_file_path):
+        elif self._add_error(Errors.pack_metadata_version_should_be_raised(self.pack, old_version),
+                             metadata_file_path):
             return False
-
-        return True
+        return False
```
Comment (dansterenson) [inline demisto_sdk/commands/common/hook_validations/pack_unique_files.py:313]: Why was this changed?

File: demisto_sdk/commands/common/errors.py [hunk 2]
```diff
@@ -19,7 +19,7 @@
     'IF100', 'IF106', 'IF115', 'IF116',
     'IN109', 'IN110', 'IN122', 'IN124', 'IN126', 'IN128', 'IN135', 'IN136', 'IN139', 'IN144', 'IN145',
     'MP106',
-    'PA113', 'PA116', 'PA124', 'PA125', 'PA127', 'PA129',
+    'PA113', 'PA116', 'PA124', 'PA125', 'PA127', 'PA129', 'PA130'
```
Comment (dansterenson) [inline demisto_sdk/commands/common/errors.py:22]: Why did you add it to the ALLOWED_IGNORE_ERRORS?

File: demisto_sdk/commands/common/tests/pack_metadata_validator_test.py [hunk 3]
```diff
@@ -149,3 +150,25 @@ def test_metadata_validator_empty_categories(self, mocker):
...
+        metadata = os.path.join(self.__class__.FILES_PATH, 'pack_metadata_invalid_format_version.json')
+        mocker.patch.object(tools, 'get_dict_from_file', return_value=({'approved_list': []}, 'json'))
+        mocker.patch.object(PackUniqueFilesValidator, '_read_file_content',
+                            return_value=TestPackMetadataValidator.read_file(metadata))
+        mocker.patch.object(PackUniqueFilesValidator, '_is_pack_file_exists', return_value=True)
+        mocker.patch.object(BaseValidator, 'check_file_flags', return_value='')
+
+        validator = PackUniqueFilesValidator('fake')
+        assert not validator.validate_pack_meta_file()
+        assert "[PA130] - Wrong version format in metadata.json change to a 0.0.0 type format" in validator.get_errors()
```
Comment (dansterenson) [inline demisto_sdk/commands/common/tests/pack_metadata_validator_test.py:174]: Does the previous tests check the scenario where the version is correct? make sure your validation runs on them.

**Review by dansterenson — CHANGES_REQUESTED** [review]
Nice!

File: demisto_sdk/commands/common/hook_validations/pack_unique_files.py (hunk 1 above)
Comment (omerKarkKatz) [reply demisto_sdk/commands/common/hook_validations/pack_unique_files.py:313]: you are right i did there some modification that was necessary in the end and this remain i will change it back. 
although it is i strange code and will be happy to go over with you.

File: demisto_sdk/commands/common/tests/pack_metadata_validator_test.py (hunk 3 above)
Comment (omerKarkKatz) [reply demisto_sdk/commands/common/tests/pack_metadata_validator_test.py:174]: didn't understand you want me to make a test of a correct version check as well ??

File: demisto_sdk/commands/common/errors.py (hunk 2 above)
Comment (omerKarkKatz) [reply demisto_sdk/commands/common/errors.py:22]: was not sure were to put it, looks like were most of the error and similar errors to mine were. 
if you don't think it's the correct location. plz tell me what you think about it.

File: demisto_sdk/commands/common/tests/pack_metadata_validator_test.py (hunk 3 above)
Comment (omerKarkKatz) [reply demisto_sdk/commands/common/tests/pack_metadata_validator_test.py:174]: Done

File: demisto_sdk/commands/common/errors.py [hunk 4]
```diff
@@ -20,7 +20,7 @@
...
```
Comment (Noy-Maimon) [inline demisto_sdk/commands/common/errors.py:23]:
```suggestion
    'PA113', 'PA116', 'PA124', 'PA125', 'PA127', 'PA129', 'PA130',
```

File: demisto_sdk/commands/common/tests/pack_metadata_validator_test.py [hunk 5]
```diff
@@ -149,3 +150,54 @@ def test_metadata_validator_empty_categories(self, mocker):
...
```
Comment (dansterenson) [inline demisto_sdk/commands/common/tests/pack_metadata_validator_test.py:174]:
```suggestion
        assert "[PA130] - Pack metadata version format is not valid. Please fill in a valid format (example: 0.0.0)" in validator.get_errors()
```

File: demisto_sdk/commands/common/tests/pack_metadata_validator_test.py [hunk 6]
```diff
@@ -149,3 +150,54 @@ def test_metadata_validator_empty_categories(self, mocker):
...
+
+    @pytest.mark.parametrize('version,expected', version_checks)
+    def test_is_version_format(self, version, expected):
+        """
+                Given:
+                    - A version to be checked by the _is_version_format function
+
+                When:
+                    - We want to check the correctness of the version while the
+                        _is_pack_meta_file_structure_valid is validating the packs meta data
```
Comment (dansterenson) [inline demisto_sdk/commands/common/tests/pack_metadata_validator_test.py:195]:
```suggestion
                    - Validating meta data structure.
```

**Review by dansterenson — CHANGES_REQUESTED** [review]

File: demisto_sdk/commands/common/hook_validations/pack_unique_files.py [hunk 7]
```diff
@@ -417,6 +415,11 @@ def _is_pack_meta_file_structure_valid(self):
                                    self.pack_meta_file):
                     return False
 
+            # check format of metadata version
+            version = metadata.get(PACK_METADATA_CURR_VERSION, '0.0.0')
+            if not self._is_version_format(version):
+                return False
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply demisto_sdk/commands/common/hook_validations/pack_unique_files.py:421]: There is a reason i did this followed a convention of a function i integrated my code to. 
further more i dont want to return on True..

File: demisto_sdk/commands/common/hook_validations/pack_unique_files.py [hunk 8]
```diff
@@ -308,10 +307,9 @@ def validate_version_bump(self):
         current_version = current_meta_file_content.get('currentVersion', '0.0.0')
         if LooseVersion(old_version) < LooseVersion(current_version):
             return True
-
-        elif self._add_error(Errors.pack_metadata_version_should_be_raised(self.pack, old_version), metadata_file_path):
+        elif self._add_error(Errors.pack_metadata_version_should_be_raised(self.pack, old_version),
+                             metadata_file_path):
```
[the remark this answers is not in the document]
Comment (omerKarkKatz) [reply demisto_sdk/commands/common/hook_validations/pack_unique_files.py:311]: Done

File: demisto_sdk/commands/common/constants.py [hunk 9]
```diff
@@ -1073,6 +1073,7 @@ class PB_Status:
 LAYOUTS_CONTAINERS_OLDEST_SUPPORTED_VERSION = '6.0.0'
 GENERIC_OBJECTS_OLDEST_SUPPORTED_VERSION = '6.5.0'
 FEATURE_BRANCHES = ['v4.5.0']
+VERSION_REGEX = r'((\d+\.){2})(\d+){1}'
```
Comment (yaakovi) [inline demisto_sdk/commands/common/constants.py:1076]:
```suggestion
VERSION_REGEX = r'(\d+\.){2}\d+'
```

File: demisto_sdk/commands/common/tests/pack_metadata_validator_test.py [hunk 10]
```diff
@@ -149,3 +150,53 @@ def test_metadata_validator_empty_categories(self, mocker):
...
+                    - A version to be checked by the _is_version_format function
+
+                When:
+                    - Validating meta data structure.
+
+                Then:
+                    - return True if the version is in the correct format and False otherwise
+
+
+        """
```
Comment (yaakovi) [inline demisto_sdk/commands/common/tests/pack_metadata_validator_test.py:200]: fix the indentations here

File: demisto_sdk/commands/common/hook_validations/pack_unique_files.py [hunk 11]
```diff
@@ -504,6 +506,21 @@ def _is_approved_usecases(self) -> bool:
                 return False
         return True
 
+    def _is_version_format(self, version: str) -> bool:
```
Comment (yaakovi) [inline demisto_sdk/commands/common/hook_validations/pack_unique_files.py:509]: Improve function name; the current one doesn't mean anything.
```suggestion
    def _is_version_format_valid(self, version: str) -> bool:
```

File: demisto_sdk/commands/common/tests/pack_metadata_validator_test.py (hunk 10 above)
Comment (omerKarkKatz) [reply demisto_sdk/commands/common/tests/pack_metadata_validator_test.py:200]: Done

**Review by dansterenson — CHANGES_REQUESTED** [review]
Please add your task description to the CHANGELOG.md file.

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,6 +1,7 @@
 # Changelog
 * Fixed an issue in the **create-id-set** command where similar items from different marketplaces were reported as duplicated.
 * Fixed typo in demisto-sdk init
+* Added a version format validation to the pack_metadata.json as part of pack_unique_files checks.
```
Comment (dansterenson) [inline CHANGELOG.md:4]:
```suggestion
* Added a validation to the **validate** command that verifies that the version in the pack_metdata file is written in the correct format.
```

**Review by dansterenson — CHANGES_REQUESTED** [review]

**Review by yaakovi — APPROVED** [review]

**Review by dansterenson — APPROVED** [review]
